atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosServerImplementation.java

```diff
@@ -0,0 +1,233 @@
+                remoteServers,
+                LEADER_NAMESPACE,
+                optionalSecurity,
+                PaxosAcceptor.class);
+        acceptors.add(paxosResource.getPaxosAcceptor(LEADER_NAMESPACE));
+
+        Map<PingableLeader, HostAndPort> otherLeaders = Leaders.generatePingables(
+                remoteServers,
+                optionalSecurity);
+
+        ExecutorService executor = Executors.newCachedThreadPool(new ThreadFactoryBuilder()
+                .setNameFormat("atlas-leaders-%d")
+                .setDaemon(true)
+                .build());
+
+        PaxosProposer proposer = PaxosProposerImpl.newProposer(
+                paxosResource.getPaxosLearner(LEADER_NAMESPACE),
+                ImmutableList.copyOf(acceptors),
+                ImmutableList.copyOf(learners),
+                getQuorumSize(configuration),
+                executor);
+
+        // Build and store a gatekeeping leader election service
+        leaderElectionService = new PaxosLeaderElectionService(
+                proposer,
+                paxosResource.getPaxosLearner(LEADER_NAMESPACE),
+                otherLeaders,
+                ImmutableList.copyOf(acceptors),
+                ImmutableList.copyOf(learners),
+                executor,
+                paxosConfiguration.pingRateMs(),
+                paxosConfiguration.randomWaitBeforeProposalMs(),
+                paxosConfiguration.leaderPingResponseWaitMs());
+        environment.jersey().register(leaderElectionService);
+    }
+
+    private static int getQuorumSize(TimeLockServerConfiguration configuration) {
+        return configuration.cluster().servers().size() / 2 + 1;
+    }
+
+    private static Optional<SSLSocketFactory> constructOptionalSslSocketFactory(
+            PaxosConfiguration configuration) {
+        return Optional.of(SslSocketFactories.createSslSocketFactory(configuration.sslConfiguration().get()));
+    }
+
+    @Override
+    public void onStop() {
+        // Nothing to do
+    }
+
+    @Override
+    public void onStartupFailure() {
+        // Nothing
+    }
+
+    @Override
+    public TimeLockServices createInvalidatingTimeLockServices(String client) {
+        TimestampService timestampService = createPaxosBackedTimestampService(client);
+        LockService lockService = AwaitingLeadershipProxy.newProxyInstance(
+                LockService.class,
+                LockServiceImpl::create,
+                leaderElectionService);
+
+        return TimeLockServices.create(timestampService, lockService);
+    }
+
+    private TimestampService createPaxosBackedTimestampService(String client) {
+        paxosResource.addClient(client);
+
+        ExecutorService executor = Executors.newCachedThreadPool(new ThreadFactoryBuilder()
+                .setNameFormat("atlas-consensus-" + client + "-%d")
+                .setDaemon(true)
+                .build());
+
+        List<PaxosAcceptor> acceptors = getNamespacedProxies(
```

Comment:
It is. I'm going to divide the existing impl of `Leaders.create()` into several smaller pieces, so that we can reuse bits of it.